js/app.js

```diff
@@ -42,121 +60,130 @@ var app = new Vue({
           })
-          .catch(onError.bind(self));
+          .catch(onError.bind(self))
       })
     },
-    fetchBuilds: function() {
-      var self = this
-      this.projects.forEach(function(p){
-        axios.get('/projects/' + p.data.id + '/repository/branches/' + p.project.branch)
-          .then(function (response) {
-            lastCommit = response.data.commit.id
-            axios.get('/projects/' + p.data.id + '/repository/commits/' + lastCommit + '/builds')
-              .then(function (response) {
-                updated = false
-
-                build = self.filterLastBuild(response.data)
-                if (!build) {
-                  return
-                }
-                startedFromNow = moment(build.started_at).fromNow()
-
-                self.builds.forEach(function(b){
-                  if (b.project == p.project.projectName && b.branch == p.project.branch) {
-                    updated = true
-
-                    b.id = build.id
-                    b.status = build.status
-                    b.started_at = startedFromNow,
-                    b.author = build.commit.author_name
-                    b.project_path = p.data.path_with_namespace
-                    b.branch = p.project.branch
-                  }
-                });
-
-                if (!updated) {
-                  self.builds.push({
-                    project: p.project.projectName,
-                    id: build.id,
-                    status: build.status,
-                    started_at: startedFromNow,
-                    author: build.commit.author_name,
-                    project_path: p.data.path_with_namespace,
-                    branch: p.project.branch
-                  })
-                }
-              })
-              .catch(onError.bind(self));
-          })
-          .catch(onError.bind(self));
-      })
+    updateBuilds: function() {
+      const self = this
+      self.onError = null
+      self.projects.forEach(function(p) {self.fetchBuild(p)})
+      self.lastRun = lastRun()
+      self.pipelines.sort(function (a, b) { return a.project.localeCompare(b.project) })
     },
+    fetchBuild: function(p) {
+      const self = this
 
-    filterLastBuild: function(builds) {
-      if (!Array.isArray(builds) || builds.length === 0) {
-        return
+      function hasPipeline(commit) {
+        return commit.data.last_pipeline !== null && commit.data.last_pipeline.id !== undefined
       }
-      return builds[0]
+
+      function isNewOrStaleProject(commit) {
+        // Either this is a fresh project or the last_pipeline did not change between retrievals (status or id)
+        return self.pipelinesMap[p.project.key] === undefined ||
+          self.pipelinesMap[p.project.key].id !== commit.data.last_pipeline.id ||
+          self.pipelinesMap[p.project.key].status !== commit.data.last_pipeline.status
+      }
+
+      axios.get('/projects/' + p.data.id + '/repository/commits/' + p.project.branch)
```

Comment:
You could replace this with the pipeline api (`/api/v4/projects/<project_id>/pipelines?ref=<branch>`)

Reply:
Hello @ggarnier no, unfortunately I can not. Your call returns something like:
```
[
  {
    "id": 146867,
    "sha": "d1e48da8f4b394ea09c02e411d41f367a25e11e4",
    "ref": "master",
    "status": "success"
  },
  {
    "id": 143913,
    "sha": "27c2bb93dcfc7361106b5f360ae1eaa9511666c7",
    "ref": "master",
    "status": "success"
  },
  {
    "id": 139846,
    "sha": "267bfa2707c6da052f506de226aa3534e777363f",
    "ref": "master",
    "status": "success"
  }
]
```
As you can see, we can not retrieve the autor or the commit message from this, these fields are only available in `/api/v4/projects/<project_id>/repository/commits/<branch|sha1>`:
```
{
  "id": "d1e48da8f4b394ea09c02e411d41f367a25e11e4",
  "short_id": "d1e48da8",
  "title": "Merge branch 'features/BAR-20783-add-maven-repo' into 'master'",
  "created_at": "2017-12-05T13:05:00.000+00:00",
  "parent_ids": [
    "27c2bb93dcfc7361106b5f360ae1eaa9511666c7",
    "6fae45922e61470b65f3035d9605dc67a264f26c"
  ],
  "message": "Merge branch 'features/BAR-20783-add-maven-repo' into 'master'\n\nBAR-20783: Add Maven plugin repo\n\nSee merge request group/group-project-config!15",
  "author_name": "Foo Bar",
  "author_email": "foo.bar@example.com",
  "authored_date": "2017-12-05T13:05:00.000+00:00",
  "committer_name": "Foo Bar",
  "committer_email": "foo.bar@example.com",
  "committed_date": "2017-12-05T13:05:00.000+00:00",
  "stats": {
    "additions": 5,
    "deletions": 0,
    "total": 5
  },
  "status": "success",
  "last_pipeline": {
    "id": 146867,
    "sha": "d1e48da8f4b394ea09c02e411d41f367a25e11e4",
    "ref": "master",
    "status": "success"
  }
}
```

And only the single pipeline instance `/api/v4/projects/<project_id>/pipelines/<pipeline_id>` holds the information about execution dates (`created_at`, `finished_at`):
```
{
  "id": 146867,
  "sha": "d1e48da8f4b394ea09c02e411d41f367a25e11e4",
  "ref": "master",
  "status": "success",
  "before_sha": "d1e48da8f4b394ea09c02e411d41f367a25e11e4",
  "tag": false,
  "yaml_errors": null,
  "user": {
    "id": 1081,
    "name": "Foo Bar
    "username": "fbar",
    "state": "active",
    "avatar_url": "https://git.example.comuploads/-/system/user/avatar/1081/avatar.png",
    "web_url": "https://git.example.com/fbar"
  },
  "created_at": "2017-12-05T13:05:02.270Z",
  "updated_at": "2017-12-05T13:08:40.511Z",
  "started_at": "2017-12-05T13:05:07.997Z",
  "finished_at": "2017-12-05T13:08:40.469Z",
  "committed_at": null,
  "duration": 211,
  "coverage": null
}
```

Reply:
@mfriedenhagen right. I was thinking about using the `user` from the single pipeline instance request, but I guess that is the user who triggered the pipeline, not necessarily the author of the last commit.

I'm having some trouble with the request to `/api/v4/projects/<project_id>/repository/commits/<branch|sha1>`. It's breaking in my tests, because the response doesn't have the `last_pipeline` field. Here's an example:

```
{
   "id":"31b2690963b61cc790ab52de9ba852ef2e85227a",
   "short_id":"31b26909",
   "title":"test commit",
   "created_at":"2017-12-06T16:39:39.000-02:00",
   "parent_ids":[
      "ee26293ee482fc5a2bb8a0c59a0c18aa51bd87da"
   ],
   "message":"test commit\n",
   "author_name":"foo",
   "author_email":"foo@example.com",
   "authored_date":"2017-12-06T16:39:39.000-02:00",
   "committer_name":"foo",
   "committer_email":"foo@example.com",
   "committed_date":"2017-12-06T16:39:39.000-02:00",
   "stats":{
      "additions":2,
      "deletions":0,
      "total":2
   },
   "status":"failed"
}
```

Reply:
Well, our inhouse GitLab CE 10.2.4 and the official one have the status:
```
curl -s -HPRIVATE-TOKEN:$GITLAB_COM_TEST_OAUTH2 https://gitlab.com/api/v4/projects/mfriedenhagen%2Fvbdoc-test/repository/commits/master | jsonpp
{
  "id": "0843d017d0936fb7ce620aada33362766c19515b",
  "short_id": "0843d017",
  "title": "Add .gitlab-ci.yml",
  "created_at": "2017-08-12T19:57:19.000+00:00",
  "parent_ids": [
    "b7f7a7ed915832fcb599b7e9d470bfba1c2bca4d"
  ],
  "message": "Add .gitlab-ci.yml",
  "author_name": "Mirko Friedenhagen",
  "author_email": "mfriedenhagen@gmail.com",
  "authored_date": "2017-08-12T19:57:19.000+00:00",
  "committer_name": "Mirko Friedenhagen",
  "committer_email": "mfriedenhagen@gmail.com",
  "committed_date": "2017-08-12T19:57:19.000+00:00",
  "stats": {
    "additions": 5,
    "deletions": 0,
    "total": 5
  },
  "status": "success",
  "last_pipeline": {
    "id": 11410130,
    "sha": "0843d017d0936fb7ce620aada33362766c19515b",
    "ref": "master",
    "status": "success"
  }
}                                                            
```

Which version do you have?

Reply:
I've tested in gitlab 9.5.9. Maybe it's because there's no successful pipeline for that project. I'll do more tests with it.

Reply:
@mfriedenhagen `last_pipeline` field was added to commits API in gitlab 10.0: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/13521, that's why it's not working for me.

To make this work in gitlab 9 and 10, we will have to make a few changes:

1. get `/projects/<id>/pipelines/?ref=<branch>` to fetch the latest pipelines
2. get `/projects/<id>/repository/commits/<commit_id>`, using the commit it from the latest pipeline